File: domain.guice/src/main/java/org/openmedstack/domain/guice/ContainerAggregateFactory.java
package org.openmedstack.domain.guice;

import com.google.inject.Injector;
import org.openmedstack.domain.Aggregate;
import org.openmedstack.domain.Memento;
import org.openmedstack.eventstore.IConstructAggregates;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;

class ContainerAggregateFactory implements IConstructAggregates {
    private final Injector _container;

    public ContainerAggregateFactory(Injector container){
        _container = container;
    }

    @Override
    public <T extends Aggregate> T build(Class<T> type, String id, Memento snapshot) {
        T result = null;
        try {
            var optional = Arrays.stream(type.getConstructors()).findFirst();
            if(optional.isPresent()){
                var constructor = optional.get();
                var parameters = Arrays.stream(constructor.getParameterTypes()).map(t->{
                    if (String.class.isAssignableFrom(t)) {
                        return id;
                    }
                    if(Memento.class.isAssignableFrom(t)){
                        return snapshot;
                    }
                    return _container.getInstance(t);
                });
                result = (T)optional.get().newInstance(parameters.toArray());
            }
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
        return result;
    }
}

File: domain.guice/src/main/kotlin/org/openmedstack/domain/guice/ContainerAggregateFactory.kt
package org.openmedstack.domain.guice

import com.google.inject.Inject
import com.google.inject.Injector
import org.openmedstack.domain.Aggregate
import org.openmedstack.domain.Memento
import org.openmedstack.eventstore.IConstructAggregates
import java.lang.reflect.InvocationTargetException
import java.util.*

internal class ContainerAggregateFactory @Inject constructor(private val _container: Injector) : IConstructAggregates {
    override inline fun <T : Aggregate> build(type: Class<T>, id: String, snapshot: Memento?): T {
        var result: T? = null
        try {
            val optional = Arrays.stream(type.constructors).findFirst()
            if (optional.isPresent) {
                val constructor = optional.get()
                val parameters = Arrays.stream(constructor.parameterTypes).map { t: Class<*>? ->
                    if (String::class.java.isAssignableFrom(t)) {
                        return@map id
                    }
                    if (Memento::class.java.isAssignableFrom(t)) {
                        return@map snapshot
                    }
                    _container.getInstance(t)
                }
                result = optional.get().newInstance(*parameters.toArray()) as T
            }
        } catch (e: InstantiationException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException(e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException(e)
        }
        return result!!
    }
}
